### ISSUE TITLE
MQTT: Status node remains online

### ISSUE THREAD

**Issue by gepd** [node-red/node-red#981]
Hi, I'm having some issues trying to getting the offline status from a node.

In a flow tab, I've a MQTT node. I can get the info sent by the node (nodemcu), but when I disconect the nodemcu, the MQTT node remains online. I had readed [here](http://nodered.org/docs/creating-nodes/status) this status, it should change if the node goes offline.

What am I doing wrong?

**Comment by hardillb** [comment]
Please use the mailing list for questions, issues are for actual defects. https://groups.google.com/forum/#!forum/node-red

The MQTT nodes have their own connection to the broker, the status represents that connection, not anything else. The topic a message is published on does not directly represent another MQTT client, if you want to track the status of MQTT clients then you will need to look at using things like the Last Will and Testament feature

**Comment by muglar** [comment]
I believe this is still an issue.  I've tested this with Node Red 15.2 connected to ActiveMQ.  The issue is the the status of the MQTT out node stays in the "connected mode" for a long time even when disconnected from the broker.  To disconnect from the broker I'm disabling the broker's NIC.  The node red client shows the "connected status" for ~16 minutes before it changes to disconnected state and then connecting state.

If you re-deploy the project it instantly shows the correct state.

As well, if you reset the broker or shutdown the broker the status changes instantaneously.

I would expect that the client would be able to determine the connection status with better resolution.  16 minutes is too long for most applications.

**Comment by dceejay** [comment]
What is your TCP timeout ?

**Comment by muglar** [comment]
Good question.  I guess it would be the default TCP timeout set in
ubuntu....I'm not too sure where this is set to check however.

On Sat, Jan 7, 2017 at 12:07 PM, Dave Conway-Jones <notifications@github.com

**Comment by muglar** [comment]
So I've tested again to get exact numbers.  The behavior is the same for two separate node-red clients.

In the one node-red client I adjusted the socketTimeout from default of no timeout to a timeout of 120 seconds.

I then disconnected the broker and time the exact time for the node-red mqtt client to show a time out.  Both nodes timed out after 15 minutes, 45 seconds.  So the TCP socketTimeout didn't have any affect....is there something else you are referring too?

Below is my exact version etc.

![image](https://cloud.githubusercontent.com/assets/17802517/21746863/1f2f8a7e-d522-11e6-8df0-24f3a29611e5.png)

**Comment by muglar** [comment]
I've done a little more searching on this and found this article.

It seems the tcp_retries2 setting in ubuntu is affecting the TCP timeout.  The default value of 15 yields a hypothetical timeout of 924.6 seconds which exactly matches my 15 minutes and 45 seconds I recorded.

The MQTT client is supposed to disconnect based on the keepalive timer which I set to 15 seconds or the system TCP connection time out (which ever happens first)  The keepalive timer is much less than the TCP timeout, so it should rule, however this doesn't seem to be the case and the TCP timeout seems to be doing its job, however with such a long delay.

I'm curious if anyone would ever adjust this tcp_retries2 parameter......?






/proc/sys/net/ipv4/tcp_retries1
/proc/sys/net/ipv4/tcp_retries2
.

tcp_retries1 - INTEGER


 This value influences the time, after which TCP decides, that
    something is wrong due to unacknowledged RTO retransmissions,
    and reports this suspicion to the network layer.
    See tcp_retries2 for more details.

    RFC 1122 recommends at least 3 retransmissions, which is the
    default.


tcp_retries2 - INTEGER

This value influences the timeout of an alive TCP connection,
when RTO retransmissions remain unacknowledged.
Given a value of N, a hypothetical TCP connection following
exponential backoff with an initial RTO of TCP_RTO_MIN would
retransmit N times before killing the connection at the (N+1)th RTO.

The default value of 15 yields a hypothetical timeout of 924.6
seconds and is a lower bound for the effective timeout.
TCP will effectively time out at the first RTO which exceeds the
hypothetical timeout.

RFC 1122 recommends at least 100 seconds for the timeout,
which corresponds to a value of at least 8.
